### DISCUSSION TITLE
Could not find gen-ctl-io program during configuration

### DISCUSSION THREAD

**Discussion by kalosu** [NanoComp/meep#2430] [Q&A] [answered]
Hello again friends from the meep community,

I am trying to install meep by building it from source based on the documentation given on https://meep.readthedocs.io/en/latest/Build_From_Source/.

I have installed all of the dependences that are mentioned there on non-standard directories (I am trying it to build meep on a HPC cluster) and by the time I want to run the autogen file for meep I get the following error message:

configure: error: could not find gen-ctl-io program; check your PATH

From what I have seen, this program is some sort of dependence that was supposed to be built while installing libctl. Notice that I have specified the directory for libctl by using the option ----with-libctl=dir. 

Any idea about what could be happening here? What could I do in order to solve this problem?

**Comment by stevengj** [comment]
If you didn't do `make install` with `libctl` then it didn't get installed.  In general I would recommend doing `make install` for all dependencies, and configure with e.g. `--prefix=$HOME/install` to install into your home directory under `~/install/bin`, `~/install/lib`, etcetera.  Then be sure to add `~/install/bin` to your `$PATH`.  See also [https://meep.readthedocs.io/en/latest/Build_From_Source/#installation-paths](Installation paths).

(However, if you are just using the Python interface, then you can build everything without the Scheme/Guile stuff like `gen-ctl-io` — just add `--without-guile` when configuring libctl, add `--without-libctl` when configuring MPB, and add `--without-scheme` when configuring Meep.)

**Comment by kalosu** [comment]
Hi, thanks for your feedback.

I have double checked and re built libctl with the option --without-guile.

Then while trying to build meep with --without-scheme I get now the following:

configure: error: libctlgeom was not found, is required for Python interface

Again, if I am not wrong this is also another thing that should be installed once I build and install libctl right?

**Comment by kalosu** [comment]
By the way, I found one article from 5 years ago in where the same issue was present.
There it was mentioned that in order to have libctlgeom.so it was needed to use --enable-shared while configuring libctl.

I have done this but nevertheless, I still get the same error message while building for meep.

Any idea about what could I do in order to check for this problem?

**Comment by kalosu** [comment]
Hi again,
I have given it another fresh test to building meep by following the build documentation.

Nevertheless, I get again the same error:

configure: error: libctlgeom was not found, is required for Python interface

While configuring meep I have done:

sh autogen.sh --enable-shared --with-mpi --with-openmp --with-libctl=$MY_OWN --without-scheme PYTHON=python3 LDFLAGS="-L${HDF5_LIB_DIR} -L${MY_OWN}" CPPFLAGS="-I${HDF5_INC_DIR} -I${MY_OWN}" 

where $MY_OWN points to the directory in where all of the installed libraries and header files are located.

I have checked under the $MY_OWM/lib folder and over there I have the following shared object file:
libctlgeom.so

Here I also include the error section of the generated config.log file while trying to configure for meep:

configure:29233: $? = 0
configure:29258: checking for swig
configure:29274: found /opt/bwhpc/common/devel/swig/4.1.0/bin/swig
configure:29286: result: swig
configure:29311: error: libctlgeom was not found, is required for Python interface

What could be potentially going on here?? Any ideas about what could I do?

**Reply by kalosu** [reply]
I forgot it: I also included the --without-scheme flag for the configuration

**Comment by stevengj** [comment]
> configure: error: libctlgeom was not found, is required for Python interface

Into what directory did you install libctl (i.e. what did you pass for `--prefix`)?   Did you check that that directory (`prefix/lib`) has `libctlgeom`?  Did you tell Meep's `configure` script where to find it using `LDFLAGS`?

**Comment by smartalecH** [comment]
The `--with-libctl` configure flag is useful here. You use it to tell meep's configure script where your libctl install lives (It should be able to find all the libctl solibs from this). 

For example, `--with-libctl=$MY_OWM/shared/libctl`

Otherwise, like Steven suggests, just add that dir to your LDFLAGS.

**Comment by kalosu** [comment, marked as answer]
Hi everyone,

I finally made it to build meep on the HPC that I have access to. It seems that I needed to add $MY_OWN/lib/ and not just $MY_OWN/ to the LDFLAGS flag.

Thanks a lot for your help!
